File: java/src/main/java/com/workshop/architecture/fitness/outside/recipient/mail/InMemoryMailSender.java
package com.workshop.architecture.fitness.outside.recipient.mail;

import com.workshop.architecture.fitness.inside.port.outbound.CustomerActivateMembershipEmail;
import com.workshop.architecture.fitness.inside.port.outbound.ForSendingEmails;
import com.workshop.architecture.fitness.infrastructure.InMemoryEmailService;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

@Component
public class InMemoryMailSender implements ForSendingEmails {

    private final InMemoryEmailService emailService;
    private final String billingSenderEmailAddress;

    public InMemoryMailSender(InMemoryEmailService emailService,
                              @Value("${workshop.billing.sender-email-address}") String billingSenderEmailAddress) {
        this.emailService = emailService;
        this.billingSenderEmailAddress = billingSenderEmailAddress;
    }

    @Override
    public void sendEmail(CustomerActivateMembershipEmail emailDetails) {
        var activationEmail = emailDetails.emailTemplate().formatted(
                        emailDetails.emailAddress(),
                        billingSenderEmailAddress,
                        emailDetails.invoiceId(),
                        emailDetails.invoiceId(),
                        emailDetails.planPrice(),
                        emailDetails.invoiceDueDate(),
                        emailDetails.invoiceId()
                )
                .replace("\n|", "\n")
                .trim();

        System.out.println(activationEmail);

        emailService.send(activationEmail);
    }
}

File: python/src/workshop_api/fitness/hexagon/outside/driven/mail/in_memory_mail_sender.py
from workshop_api.fitness.hexagon.inside.port.outbound.customer_activate_membership_email import (
    CustomerActivateMembershipEmail,
)
from workshop_api.fitness.hexagon.inside.port.outbound.for_sending_emails import (
    ForSendingEmails,
)
from workshop_api.fitness.layered.infrastructure.in_memory_email_service import (
    InMemoryEmailService,
)


class InMemoryMailSender(ForSendingEmails):
    def __init__(
        self,
        email_service: InMemoryEmailService,
        billing_sender_email_address: str,
    ) -> None:
        self.email_service = email_service
        self.billing_sender_email_address = billing_sender_email_address

    def send_email(self, email_details: CustomerActivateMembershipEmail) -> None:
        activation_email = (
            email_details.email_template.format(
                to=email_details.email_address,
                sender=self.billing_sender_email_address,
                invoice_id=email_details.invoice_id,
                amount=email_details.plan_price,
                due_date=email_details.invoice_due_date,
            )
            .replace("\n|", "\n")
            .strip()
        )

        print(activation_email)
        self.email_service.send(activation_email)
